presto-main/src/main/java/com/facebook/presto/operator/SpatialLookupJoinOperator.java

```diff
@@ -0,0 +1,259 @@
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final List<Type> probeTypes;
+    private final List<Integer> probeOutputChannels;
+    private final int probeGeometryChannel;
+    private final ListenableFuture<PagesSpatialIndex> pagesSpatialIndexFuture;
+    private final BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest;
+    private final PageBuilder pageBuilder;
+    private Page probe;
+    private int probePosition;
+    private List<Long> joinAddresses;
+    private int joinAddressIndex;
+    private Page outputPage;
+    private boolean finishing;
+    private boolean finished;
+
+    public SpatialLookupJoinOperator(
+            OperatorContext operatorContext,
+            List<Type> outputTypes,
+            List<Type> probeTypes,
+            List<Integer> probeOutputChannels,
+            int probeGeometryChannel,
+            ListenableFuture<PagesSpatialIndex> pagesSpatialIndexFuture,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest)
+    {
+        this.operatorContext = operatorContext;
+        this.probeTypes = ImmutableList.copyOf(probeTypes);
+        this.probeOutputChannels = ImmutableList.copyOf(probeOutputChannels);
+        this.probeGeometryChannel = probeGeometryChannel;
+        this.pagesSpatialIndexFuture = pagesSpatialIndexFuture;
+        this.pageBuilder = new PageBuilder(outputTypes);
+        this.spatialRelationshipTest = spatialRelationshipTest;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return probeTypes;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return pagesSpatialIndexFuture.isDone() && outputPage == null && probe == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        Verify.verify(probe == null);
+        probe = page;
+        probePosition = 0;
+
+        joinAddresses = null;
+        joinAddressIndex = 0;
+
+        processProbe();
+        tryBuildOutputPage();
+    }
+
+    private void tryBuildOutputPage()
+    {
+        if (pageBuilder.isFull()) {
+            outputPage = pageBuilder.build();
+            pageBuilder.reset();
+        }
+    }
+
+    private void processProbe()
+    {
+        Verify.verify(probe != null);
+
+        PagesSpatialIndex pagesSpatialIndex = getDone(pagesSpatialIndexFuture);
+        for (int position = probePosition; position < probe.getPositionCount(); position++) {
+            if (this.joinAddresses == null) {
+                joinAddresses = pagesSpatialIndex.findJoinAddresses(position, probe, probeGeometryChannel, spatialRelationshipTest);
+                joinAddressIndex = -1;
+            }
+
+            for (int joinAddressIndex = this.joinAddressIndex + 1; joinAddressIndex < joinAddresses.size(); joinAddressIndex++) {
+                if (pageBuilder.isFull()) {
```

Comment:
@mbasmanova you could check example code in `LookupJoinOperator`. The states (in this case `position`, etc) need to be remembered in case the operator yields and comes back again to pick up the progress. `testYield()` in `TestHashJoinOperator` sets up a framework of testing the yield code path.